frame/display_server.py

```diff
@@ -0,0 +1,73 @@
+import inotify.adapters
+import os
+import time
+
+from program import Program, Command
+
+display_appear_timeout = 10
+min_frame_run_time = 0.1
+
+def clear_wayland_display(runtime_dir: str, name: str) -> None:
+    # Clear out any existing display before waiting for the new one
+    for path in [
+        os.path.join(runtime_dir, name),
+        os.path.join(runtime_dir, name + '.lock')
+    ]:
+        if os.path.exists(path):
+            os.remove(path)
+
+def wait_for_wayland_display(runtime_dir: str, name: str) -> None:
+    # Set up inotify
+    i = inotify.adapters.Inotify()
+    i.add_watch(runtime_dir, mask=inotify.constants.IN_CREATE)
+    # Check if display has already appeared Since we've already created the watch it might be seen
+    # by both this check and inotify, but there's no way it's seen by neither.
+    if os.path.exists(os.path.join(runtime_dir, name)):
+        return
+    # Wait for display to appear
+    for event in i.event_gen(timeout_s=display_appear_timeout, yield_nones=False):
+        (_, type_names, path, filename) = event
+        if filename == name:
+            return
+    # Raise timeout error if we didn't return
+    raise RuntimeError('Wayland display ' + name + ' did not appear')
+
+class DisplayServer:
+    def __init__(self, command: Command) -> None:
+        self.command = command
+
+    def program(self, command: Command) -> Program:
+        program = Program(command, env={
+            'DISPLAY': 'no',
+            'QT_QPA_PLATFORM': 'wayland',
+        })
+        self.programs.append(program)
+        return program
+
+    def __enter__(self) -> 'DisplayServer':
+        wayland_display = 'wayland-99'
```

Comment:
Don't we want to find an available name instead?

Reply:
Changed to use the current process's PID with a fixed prefix. This avoids the race conditions we might run into if we look for a free display.